### ISSUE TITLE
Build fails with "node-gyp" errors after upgrading from vs2017-win2016 image to windows-2019

### ISSUE THREAD

**Issue by marvinpjr** [actions/virtual-environments#4898]
### Description

Submitting issue as requested [here](https://github.com/actions/virtual-environments/issues/4312#issuecomment-1010212537).

Our build works using the vs2017-win2016 image, but starting with windows-2019 (and also with windows-2022), it fails on this step:

steps:
- task: Npm@1
  displayName: 'NPM Install Grunt'
  inputs:
    command: custom
    workingDir: MyCompany.MyDept.WebApp
    verbose: false
    customCommand: 'install grunt --save-dev'

The log for that step shows this error:

2022-01-11T16:33:15.9639153Z 5233 error gyp ERR! UNCAUGHT EXCEPTION
2022-01-11T16:33:15.9639561Z 5233 error gyp ERR! stack Error: spawn C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\15.0\Bin\MSBuild.exe ENOENT
2022-01-11T16:33:15.9640087Z 5233 error gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
2022-01-11T16:33:15.9640561Z 5233 error gyp ERR! stack     at onErrorNT (node:internal/child_process:477:16)
2022-01-11T16:33:15.9641014Z 5233 error gyp ERR! stack     at processTicksAndRejections (node:internal/process/task_queues:83:21)
2022-01-11T16:33:15.9641410Z 5233 error gyp ERR! System Windows_NT 10.0.17763
2022-01-11T16:33:15.9641996Z 5233 error gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "D:\\a\\1\\s\\MyCompany.MyDept.WebApp\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_library="
2022-01-11T16:33:15.9642659Z 5233 error gyp ERR! cwd D:\a\1\s\MyCompany.MyDept.WebApp\node_modules\node-sass
2022-01-11T16:33:15.9643000Z 5233 error gyp ERR! node -v v16.13.1
2022-01-11T16:33:15.9643273Z 5233 error gyp ERR! node-gyp -v v3.8.0
2022-01-11T16:33:15.9644058Z 5233 error gyp ERR! This is a bug in `node-gyp`.
2022-01-11T16:33:15.9644423Z 5233 error gyp ERR! Try to update node-gyp and file an Issue if it does not help:
2022-01-11T16:33:15.9644842Z 5233 error gyp ERR!     <https://github.com/nodejs/node-gyp/issues>
2022-01-11T16:33:15.9645180Z 5233 error Build failed with error code: 7

I did find that the error resolved by making these changes to our package.json:	
Changing this:
    "gulp-sass": "^3.1.0",
To this:
    "gulp-sass": "^5.1.0",

And adding this:
    "sass": "^1.48.0"

I believe this change removed gulp-sass dependency on node-sass which is deprecated. See [this](https://github.com/nodejs/node-gyp/issues/2581#issue-1089104504).

I do have concerns about the implications of this change for the code base as it's rather large, and I don't really know what might be affected by it. If there's a more surgical solution, I'd love to know it. Microsoft support is also working with me to find a solution.

### Virtual environments affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [X] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

Virtual Environment > Version: 20211219.1, OS: Microsoft Windows Server 2019

### Is it regression?

20211219.1, OS Microsoft Windows Server 2016

### Expected behavior

The task succeeds. On the WS 2016 image, it works.

### Actual behavior

The step fails with the "node-gyp" error included above.

### Repro steps

I'm just running our build. Same as I was before when it worked. No changes to the pipeline configuration other than changing the image from WS2016 to WS2019 or WS2022

**Comment by nikolai-frolov** [comment]
Hello @marvinpjr. Thanks for the issue reporting. Could you please provide link to the build and/or repro steps with a simple project for investigation?

**Comment by pflajszer** [comment]
This is interesting. I've sat down on this for nearly 3 days and tried everything. What I found out by running `ng serve` on another project was this:

```powershell
C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp>ng serve
Your global Angular CLI version (13.1.3) is greater than your local version (7.3.10). The local Angular CLI version is used.

To disable this warning use "ng config -g cli.warnings.versionMismatch false".
(node:26344) [DEP0111] DeprecationWarning: Access to process.binding('http_parser') is deprecated.
(Use `node --trace-deprecation ...` to show where the warning was created)
** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **

Date: 2022-01-19T13:52:59.462Z
Hash: b82a3d1fc2c5a9c79438
Time: 25711ms
chunk {main} main.js, main.js.map (main) 908 kB [initial] [rendered]
chunk {polyfills} polyfills.js, polyfills.js.map (polyfills) 237 kB [initial] [rendered]
chunk {runtime} runtime.js, runtime.js.map (runtime) 6.08 kB [entry] [rendered]
chunk {scripts} scripts.js, scripts.js.map (scripts) 161 kB [entry] [rendered]
chunk {styles} styles.css, styles.css.map (styles) 166 kB [initial] [rendered]
chunk {vendor} vendor.js, vendor.js.map (vendor) 8.24 MB [initial] [rendered]

ERROR in ./src/theme.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/sass-loader/lib/loader.js):
Error: Node Sass does not yet support your current environment: Windows 64-bit with Unsupported runtime (93)
For more information on which environments are supported please see:
https://github.com/sass/node-sass/releases/tag/v4.14.1
    at module.exports (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\node-sass\lib\binding.js:13:13)
    at Object.<anonymous> (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\node-sass\lib\index.js:14:35)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.sassLoader (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\sass-loader\lib\loader.js:46:72)
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\webpack\lib\NormalModule.js:301:20
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:367:11
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:233:18
    at runSyncOrAsync (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:143:3)
    at iterateNormalLoaders (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:232:2)
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:205:4
    at VirtualFileSystemDecorator.readFile (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\@ngtools\webpack\src\virtual_file_system_decorator.js:39:13)
    at processResource (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:202:11)
    at iteratePitchingLoaders (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:158:10)
    at iteratePitchingLoaders (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:165:10)
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:176:18
    at loadLoader (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\loadLoader.js:47:3)
    at iteratePitchingLoaders (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:169:2)
    at iteratePitchingLoaders (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:165:10)
    at C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\LoaderRunner.js:176:18
    at loadLoader (C:\Users\PawelFlajszer\source\repos\AuditLabTwo\AuditLab.UI\ClientApp\node_modules\loader-runner\lib\loadLoader.js:47:3)
i ｢wdm｣: Failed to compile.
```

Please note the error message:
```
Error: Node Sass does not yet support your current environment: Windows 64-bit with Unsupported runtime (93)
```

On further investigation you can read in node-sass [docs](https://github.com/sass/node-sass/releases/tag/v4.14.1) for the release I'm trying to install `v4.14.1`:

<html>
<body>
<h2 style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.5em; font-weight: 600; line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid var(--color-border-muted); color: rgb(173, 186, 199); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(34, 39, 46); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Supported Environments</h2>

OS | Architecture | Node
-- | -- | --
Windows | x86 & x64 | 0.10, 0.12, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14


</body>
</html>

I was running `node v16.x.x`
Steps to fix are then:
1. Uninstall `node.js`
2. Download `node v14.15.0` from [here](https://nodejs.org/download/release/v14.15.0/). for me it's `node-v14.15.0-x64.msi` (tested this one and it works, not sure about other versions)

**Comment by marvinpjr** [comment]
Thanks @pflajszer. I think this is related to what you are saying. I did find [this post](https://stackoverflow.com/a/70213915/157480) on stack overflow, so I tried using the Node Tool step as it suggests, placing it before any NPM task steps in my pipeline. The yaml for the Node Tool step in my pipeline looks like this:

steps:
- task: NodeTool@0
  displayName: 'Use Node 14.x'
  inputs:
    versionSpec: 14.x

Unfortunately, that did not resolve the issue for me.
